Lots of (re-)implementations to support a horizontal scroller.

alpyre opened the pull request:
- Lots of (re-)implementations to support a horizontal scroller in NoWrapMode.
- TextFitNew() and TextLengthNew() are re-implemented to support very large 
input/return values and instead of taking a dummy TextExtent, TextFitNew() now
takes a TextExtentNew structure which uses LONG type members and fills it properly.
So that its te.Width member can be used instead of calling another TextLengthNew().
This provides very good optimization.
- Cursor being drawn out of gadget boundaries issue when in NoWrapMode resolved.
- Issue of highlighted text changing color when edited resolved.
- Experimental MUIA_TextEditor_HorizontalScroll now acts as a switch for turning off
   horizontal scrolling even in NoWrapMode.
- Behaviour of FLG_HScroll re-implemented (now keeps the state of 
   MUIA_TextEditor_HorizontalScroll attribute).
- New public attribute MUIA_TextEditor_HorizontalScroller.

jens-maus commented:
Thanks again for this nice pull request. Here are some comments/requests however to slightly change your source to improve certain things:
1. Please remove all your `// Alpyre` type of comments. this is what a revision control is all about and with a simple "git blame" anyone can find out who last changed a line or not. So these comments are not required and obfuscate the code only. So please remove them.
2. Please explain the difference between `MUIA_TextEditor_HorizontalScroll`and `MUIA_TextEditor_HorizontalScroller`. Why are two attributes required? and why have you defined `MUIA_TextEditor_HorizontalScroll` to be 0x2d? In addition, the naming of these new attributes seem to be somewhat awkward. A better naming would be `MUIA_TextEditor_HorizontalScrolling` and `MUIA_TextEditor_HorizontalScrollbar`
3. Please use `_mwidth()` instead of saving the value in a variable throughout your changes. _mwidth() is a macro that should always be used. Please also do so for all the other `_XXXX()` MUI macros in your changes.
4. Please use proper source code indentation and keep the original code style preserved. That means, e.g. that you shouldn't write `if (blah)` but `if(blah)` without any space between `if`and `(`. Also you should put variable declarations on the next line after `{` and not doing something like `{ LONG prop_entries;` like in line 935.
5. line 982: please put '{' on the next line and please check all your changes for similar situations.
6. line 565+566: what does this "quick fix" actually do? What happens if you remove these lines again?

So please adapt your changes according to these comments and then I will continue reviewing your changes and give some more comments before I will integrate/merge your pull request.

alpyre commented:
You're wellcome.

1) Done. (all "// Alpyre" comments removed.)
2) Firstly **MUIA_TextEditor_HorizontalScroll** was previously there in the code with the value **0x2d**. I think one of the first coders on the project (probably Allan Odgaard) put it there with the idea to implement horizontal scrolling later. All that it actually does is to set/get the FLG_HScroll flag to/from data->flags. Which then is used **only** for querying some conditions which could be respectively achieved by querying MUIA_TextEditor_WrapMode for NoWrap.
So replaced those flag queries with MUIA_TextEditor_WrapMode queries.

alpyre commented:
You're wellcome.

1) Done. (all **"// Alpyre"** comments removed.)

2) Firstly **MUIA_TextEditor_HorizontalScroll** was previously there in the code with the value **0x2d**. I think one of the first coders on the project (probably Allan Odgaard) put it there with the idea to implement horizontal scrolling later. All that it actually does is to set/get the **FLG_HScroll** flag to/from **data->flags**. This then is used **only** for querying some conditions which could be respectively achieved by querying **MUIA_TextEditor_WrapMode** for **NoWrap**.
So I replaced those flag queries with **MUIA_TextEditor_WrapMode** queries, put the **FLG_HScroll** into new use (which is to switch horizontal scrolling on/off when in NoWrapMode).
Your idea to name the attributes as **MUIA_TextEditor_HorizontalScrolling** and **MUIA_TextEditor_HorizontalScrollBar** is quite reasonable. So I did it.

3) Well, I understand you. It may be a rule to do it that way, but I have to insist that this could provide great speed optimization if those macros are evaluating to functions (ie. to a get() function) and if not it wouldn’t be that much a loss.
Btw. I changed them to be _const_ variables now which makes sense.

4) All Done.

5) Done.

6) If your remove them, the text color used drawing the cursor on highlighted text does not match the highlight color. This was a long existing bug. With this fix, it is resolved. It is a “bad quick fix” because it would be better if it was fixed inside **SetColor()** function. I just copy pasted some part the highlight implementation used in **PrintLine()** function. (This may still be the only way to fix it. To figure out we should first examine and understand how SetColor() function works which I didn’t yet).

jens-maus commented:
@alpyre 

First of all, thanks for adapting your pull request. It is closely getting into a shape that I can start accepting it soon.

Here are some more questions/remarks for you:
1. Thanks for deleting the `// Alpyre` comments.
2. Honestly, I still don't understand why you still require two different attributes for horizontal scrolling. While I understand the meaning for `MUIA_TextEditor_HorizontalScrolling` to simply turn it on/off I still don't get what the second attribute is all about?
3. Regarding the `_mwidth()` type of macros I still have to instist you still please change them back to direct use where appropriate. Obviously you are missing that they are really just macros, See here: https://github.com/amiga-mui/texteditor/blob/master/include/libraries/mui.h#L3499 And then even if they weren't macros the speed improvement isn't really so high on modern amiga systems and we are not considering real OS3/m68k hardware to be the target platform for TextEditor.mcc anymore.
4. Thanks
5. Thanks
6. If so, then please open a dedicated ticket (https://github.com/amiga-mui/texteditor/issues) for this type of problem and reference the relevant problem there please so that we don't forget about it.
7. line 936 ff.: Please fix source code indentation
8. Please explain the meaning of the `MUIA_TextEditor_HScroller_XXX` private attributes you added.
9. line 361: please remove all spaces at the end of the line
10. line 463: remove empty line
11. line 704: remove all spaces at the end of the line
12. line 435 ff.: please fix source code indentation style.
13. line 828+894: add space before `return()`
14. line 340: add separate value definitions for all the LONG values and don't use "," to concatenate definition after another. This is bad coding style. Thus, each LONG value on a separate line starting with "LONG".

And please make sure to not add further changes to your master branch that are not related to the horizontal scrollbar change proposed here. Every change of functionality should be put into separate pull requests instead referring different git branches.

jens-maus commented on mcc/Dispatcher.c:
```diff
@@ -489,6 +497,18 @@ static IPTR mShow(struct IClass *cl, Object *obj, Msg msg)
     MUIA_TextEditor_Prop_Visible,     data->maxlines*data->fontheight,
     TAG_DONE);
 
+  {
```
Please remove the '{' and '}' here because they are not required.

jens-maus commented on mcc/GetSetAttrs.c:
```diff
@@ -347,7 +358,7 @@ IPTR mSet(struct IClass *cl, Object *obj, struct opSet *msg)
       break;
 
-      case MUIA_TextEditor_Prop_Release:
+      case MUIA_TextEditor_Prop_Release:                                                               
```
Remove unrequired spaces at the end of the line here

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -426,11 +426,19 @@ void SetCursor(struct InstData *data, LONG x, struct line_node *line, BOOL Set)
   UWORD flow;
 
+  if(data->rgbMode == TRUE)
+  { highlightColor.isRGB = TRUE;
```
fix indentation here: `highlightColor.isRGB` on a new line.

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -586,10 +598,9 @@ void SetCursor(struct InstData *data, LONG x, struct line_node *line, BOOL Set)
-
-      if(clipping)
-        RemoveClipping(data);
     }
```
add an empty line between '}' and the following `if(clipping)`

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -779,3 +790,106 @@ LONG CountLines(struct InstData *data, struct MinList *lines)
+    line_nr++;
+  }
+
+return(result + cursor_width);
```
add space before `return()`

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -779,3 +790,106 @@ LONG CountLines(struct InstData *data, struct MinList *lines)
+      oldline_nr++;
+    }
+
+return (oldline_nr < data->visual_y || oldline_nr >= data->visual_y + data->maxlines);
```
add space before `return()`

alpyre commented:
2) There is only one new public attribute which is (with its new name) **MUIA_TextEditor_HorizontalScrollBar**. Which is a **"Set Only"** attribute to connect the gadget to a Horizontal Scrollbar object (similar to the MUIA_TextEditor_Slider). 
The other one (MUIA_TextEditor_HorizontalScroll - now renamed as **MUIA_TextEditor_HorizontalScrolling** **and made public**) was already in the code which was implemented **only** to have the same functionality with **MUIA_TextEditor_WrapMode** (which was something quite unnecessary).
Now I put it into use for switching horizontal scrolling on/off.
For example: a developer may want a texteditor that has **NoWrapping** and also does not scroll horizontally either. Such a gadget can be set with this attribute.
(It is switched on by default.)

3) I am convinced. Done.

6) Done.

7) Done.

8) - **MUIA_TextEditor_HScroller_Pos** is **xpos** of gadget. (its notifications are connected to the **MUIA_Prop_First** of the scroller.)
- **MUIA_TextEditor_HScroller_Vis** is **_mwidth(object)**. (its notifications are connected to the **MUIA_Prop_Visible**)
- **MUIA_TextEditor_HScroller_Ent** is the current longest line (in pixel length). (its notifications are connected to the **MUIA_Prop_Entries**)

9) Done
10) Done
11) Done
12) Done.
13) Done.
14) Done.
Allright.

jens-maus commented on include/mui/TextEditor_mcc.h:
```diff
@@ -109,6 +109,8 @@ extern "C" {
 #define MUIA_TextEditor_MatchedKeyword        (TextEditor_Dummy + 0x41)
 #define MUIA_TextEditor_CursorIndex           (TextEditor_Dummy + 0x42)
 #define MUIA_TextEditor_RGBMode               (TextEditor_Dummy + 0x45)
+#define MUIA_TextEditor_HorizontalScrolling   (TextEditor_Dummy + 0x2d)
```
After thinking about these two new attributes again please do:
1. Remove `MUIA_TextEditor_HorizontalScrolling` completly and just set FLG_HScroll if `MUIA_TextEditor_HorizontalScollbar` is set to non-NULL. Otherwise remove the flag. Thus, by passing a NULL pointer to that attribute it will automatically disable Horizontal scrolling. So there is no need for another BOOLEAN attribute like that one.
2. Rename `MUIA_TextEditor_HorizontalScrollbar` to `MUIA_TextEditor_HorizontalSlider` because TextEditor already uses the term "Slider" for the vertical slider gadget

jens-maus commented on mcc/CaseConversion.c:
```diff
@@ -81,7 +81,8 @@ static void MangleCharacters(struct InstData *data, char (*change)(char c))
   }
 
-  data->HasChanged = TRUE;
+  data->HasChanged  = TRUE;
```
revert the added space before the `= TRUE;` part (also in other source code areas where you changed the data->HasChanged indentation. This is an unnecessary change.

jens-maus commented on mcc/Dispatcher.c:
```diff
@@ -159,6 +164,7 @@ static IPTR mNew(struct IClass *cl, Object *obj, struct opSet *msg)
             setFlag(data->flags, FLG_ActiveOnClick);
             setFlag(data->flags, FLG_PasteStyles);
             setFlag(data->flags, FLG_PasteColors);
+            setFlag(data->flags, FLG_HScroll);
```
Is this really correct that you enable horizontal scrolling on per default? I think this should be dependent on the fact if a developer specified a horizontal slider upon object creation.

alpyre replied:
Imagine a TextEditor gadget created without setting a vertical slider to it. Does it scroll vertically with keyboard? It does.
I think we should have the same approach for horizontal scrolling as well. So it should be on by default.
Being able to turn off the horizontal scroll action by setting a NULL for its HorizontalSlider attribute is quite a good idea indeed.

jens-maus replied:
Ok, you are right, so keep that setFlag() here and we will see what effect this might have.

jens-maus commented on mcc/Dispatcher.c:
```diff
@@ -190,6 +196,7 @@ static IPTR mNew(struct IClass *cl, Object *obj, struct opSet *msg)
             // they were set during OM_NEW
             mSet(cl, obj, (struct opSet *)msg);
             data->visual_y = 1;
+            data->xpos     = 0;
```
remove added spaces before `=` - no need to indent to the same level like the line before.

jens-maus commented on mcc/EditorStuff.c:
```diff
@@ -485,7 +485,8 @@ BOOL MergeLines(struct InstData *data, struct line_node *line)
   next = GetNextLine(line);
 
-  data->HasChanged = TRUE;
+  data->HasChanged  = TRUE;
```
indentation

jens-maus commented on mcc/EditorStuff.c:
```diff
@@ -785,7 +786,8 @@ BOOL SplitLine(struct InstData *data, LONG x, struct line_node *line, BOOL move_
     struct Grow newColorGrow;
 
-    data->HasChanged = TRUE;
+    data->HasChanged  = TRUE;
```
indentation

jens-maus commented on mcc/EditorStuff.c:
```diff
@@ -1271,7 +1273,8 @@ static void UpdateChange(struct InstData *data, LONG x, struct line_node *line,
     OptimizedPrint(data, skip, line, line_nr, width);
     ScrollIntoDisplay(data);
-    data->HasChanged = TRUE;
+    data->HasChanged  = TRUE;
```
indentation

jens-maus commented on mcc/GetSetAttrs.c:
```diff
@@ -308,6 +317,8 @@ IPTR mSet(struct IClass *cl, Object *obj, struct opSet *msg)
         // make sure a possibly existing slider is disabled as well
         if(data->slider != NULL)
           set(data->slider, MUIA_Disabled, ti_Data);
+        if(data->hscroller != NULL)
```
please use `data->hslider` instead of hscroller to keep same terminus.

jens-maus commented on mcc/GetSetAttrs.c:
```diff
@@ -556,7 +583,8 @@ IPTR mSet(struct IClass *cl, Object *obj, struct opSet *msg)
           data->Pen = pen;
           AddColor(data, &data->blockinfo, &data->Pen);
-          data->HasChanged = TRUE;
+          data->HasChanged  = TRUE;
```
indentation

jens-maus commented on mcc/GetSetAttrs.c:
```diff
@@ -734,7 +805,8 @@ IPTR mSet(struct IClass *cl, Object *obj, struct opSet *msg)
             lines = data->maxlines-(start-1);
           DumpText(data, data->visual_y+start-1, start-1, start-1+lines, TRUE);
-          data->HasChanged = TRUE;
+          data->HasChanged  = TRUE;
```
indentation

jens-maus commented on mcc/TextEditor-Test.c:
```diff
@@ -176,7 +176,7 @@ static const char *page_titles[] = { "Shown", "Hidden", NULL };
 int main(void)
 {
-  void    *slider;
+  void    *slider, *hscroller;
```
use `hslider` instead of `hscroller`

jens-maus commented on mcc/UndoFunctions.c:
```diff
@@ -204,7 +204,9 @@ BOOL Undo(struct InstData *data)
 
       if(isFlagClear(data->flags, FLG_UndoLost))
-        data->HasChanged = FALSE;
+      { data->HasChanged  = FALSE;
```
indentation

jens-maus commented:
In addition to the last bunch of requested changes above please also make sure to adapt the Autodocs (https://github.com/amiga-mui/texteditor/blob/master/doc/MCC_TextEditor.doc) and please add the new attributes you have created there with a proper developer documentation.

alpyre commented:
Would you like me to also change the **private** attributes MUIA_TextEditor_HScroller_XXX as ..._HSlider_XXX ?

jens-maus commented:
Yes, you can also change HScroller to HSlider

alpyre commented:
Good morning.
I belive we are there at last. I've made the required additions to:
- ChangeLog
- doc/MCC_TextEditor.doc
- doc/MCC_TextEditor.readme

I hope it is all good to go.

jens-maus commented on mcc/Dispatcher.c:
```diff
@@ -898,6 +920,30 @@ DISPATCHER(_Dispatcher)
+        if(data->CPos_X != oldx)
+          ScrollIntoView(data);
+
+        if (data->hslider)
```
remove space between `if` and `(`

jens-maus commented on mcc/Dispatcher.c:
```diff
@@ -898,6 +920,30 @@ DISPATCHER(_Dispatcher)
+          LONG prop_entries;
+          get(data->hslider, MUIA_Prop_Entries, &prop_entries);
+
+          if (prop_entries != data->longestline)
```
remove space between `if` and `(`

jens-maus commented on mcc/GetSetAttrs.c:
```diff
@@ -444,6 +452,26 @@ IPTR mSet(struct IClass *cl, Object *obj, struct opSet *msg)
+      case MUIA_TextEditor_HSlider_Pos:
+      {
+        // If the gadget is NOT in NoWrapMode ignore this.
+        if (data->WrapMode != MUIV_TextEditor_WrapMode_NoWrap)
```
remove space between `if` and `(`

jens-maus commented on mcc/GetSetAttrs.c:
```diff
@@ -598,6 +627,52 @@ IPTR mSet(struct IClass *cl, Object *obj, struct opSet *msg)
+      case MUIA_TextEditor_HorizontalSlider:
+      {
+        if(ti_Data)
+	{
```
incorrect indentation. bring { to the same level like if()

jens-maus commented on mcc/GetSetAttrs.c:
```diff
@@ -598,6 +627,52 @@ IPTR mSet(struct IClass *cl, Object *obj, struct opSet *msg)
+                data->hslider, 3, MUIM_NoNotifySet, MUIA_Prop_DeltaFactor, MUIV_TriggerValue);
+          }
+        }
+	else
```
else block has incorrect indentation. add space to move else block one space forward.

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -543,15 +561,11 @@ void SetCursor(struct InstData *data, LONG x, struct line_node *line, BOOL Set)
       for(c = start; c <= stop; c++)
       {
         SetColor(data, rp, &colors[1+c], line->line.Highlight, TRUE);
+        if (line->line.Highlight)
```
remove space between `if` and `(`

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -779,3 +795,106 @@ LONG CountLines(struct InstData *data, struct MinList *lines)
+  /* Since this function will be only used in NoWrapMode and
+     in that mode line_nr's will always match the real lines
+     we can find the top (real) line just by traversing the list */
+
```
remove empty line

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -779,3 +795,106 @@ LONG CountLines(struct InstData *data, struct MinList *lines)
+
+  /* We found the top line displayed. Now we'll traverse until the
+     last displayed line meanwhile calculating their pixel lengths */
+
```
remove empty line

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -779,3 +795,106 @@ LONG CountLines(struct InstData *data, struct MinList *lines)
+  LONG xmin = 0;
+  LONG xmax;
+
+  if (isFlagClear(data->flags, FLG_HScroll))
```
remove space between `if` and `(`

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -779,3 +795,106 @@ LONG CountLines(struct InstData *data, struct MinList *lines)
+
+  xmax = xmin + cursor_width;
+
+  if (xmin < (LONG)data->xpos)
```
remove space between `if` and `(`

jens-maus commented on mcc/MixedFunctions.c:
```diff
@@ -779,3 +795,106 @@ LONG CountLines(struct InstData *data, struct MinList *lines)
+ *---------------------------------------------------------------*/
+BOOL MovedIntoDisplay(struct InstData *data, struct line_node *oldline)
+{
+    struct line_node *line = GetFirstLine(&data->linelist);
```
wrong indentation until return(). remove one space in each line to bring to same level like return()

alpyre replied:
This function is never used (it was kind'a experimental). I'm completely removing it.

jens-maus commented on mcc/Navigation.c:
```diff
@@ -34,12 +34,17 @@ struct pos_info pos;
 LONG FlowSpace(struct InstData *data, UWORD flow, STRPTR text)
 {
   LONG flowspace = 0;
+  LONG textlength = TextLengthNew(&data->tmprp, text, LineCharsWidth(data, text)-1, data->TabSizePixels);
```
just put `LONG textlength = 0;` variable definition here. and put assignment to TextLengthNew() after `ENTER()` debug statement or otherwise if TextLengthNew() crashes we won't be able to identify the crash location easily.

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -45,7 +52,22 @@ LONG TextLengthNew(struct RastPort *rp, const char *string, ULONG count, LONG ta
       if(count0 != 0)
-        result += TextLength(rp, tptr0, count0);
+      {
+        if (count0 <= C_LIMIT)
```
remove space between `if` and `(`

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -45,7 +52,22 @@ LONG TextLengthNew(struct RastPort *rp, const char *string, ULONG count, LONG ta
+            l_count -= C_LIMIT;
+            tptr0 += C_LIMIT;
+          }
+          result += TextLength(rp, tptr0, l_count);
```
add empty line after `{` and before `result +=...`

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -55,7 +77,22 @@ LONG TextLengthNew(struct RastPort *rp, const char *string, ULONG count, LONG ta
       if(count0 != 0)
-        result += TextLength(rp, tptr0, count0);
+      {
+        if (count0 <= C_LIMIT)
```
remove space between `if` and `(`

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -93,25 +133,104 @@ ULONG TextFitNew(struct RastPort *rp, const char *string, ULONG strLen, struct T
       // add the number of fitting remaining characters
-      result += TextFit(rp, tptr0, strLen0, textExtent, constrainingExtent, strDirection, constrainingBitWidth, constrainingBitHeight);
+
+      if (constrainingBitWidth <= SHRT_MAX)
```
remove space between `if` and `(`

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -76,12 +113,15 @@ LONG TextLengthNew(struct RastPort *rp, const char *string, ULONG count, LONG ta
   const char *tptr0 = string;
+  struct TextExtent textExtent;
+  ULONG te_width  = 0;
+  ULONG te_height = constrainingBitHeight < rp->TxHeight ? 0 : rp->TxHeight;
```
move `te_height = constraining....` assignment after `ENTER()` and just leave variable definition here.

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -93,25 +133,104 @@ ULONG TextFitNew(struct RastPort *rp, const char *string, ULONG strLen, struct T
+
+      if (constrainingBitWidth <= SHRT_MAX)
+      {
+        if (strLen0 != 0)
```
remove space between `if` and `(`

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -93,25 +133,104 @@ ULONG TextFitNew(struct RastPort *rp, const char *string, ULONG strLen, struct T
+          tptr0   += l_chars;
+          result  += l_chars;
+        }
+        if (strLen0 > 0)
```
add empty line between `}` and `if()`

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -93,25 +133,104 @@ ULONG TextFitNew(struct RastPort *rp, const char *string, ULONG strLen, struct T
       // we found a TAB, calculate the number of fitting characters so far
-      if(strLen0 != 0)
-        result += TextFit(rp, tptr0, strLen0, textExtent, constrainingExtent, strDirection, constrainingBitWidth, constrainingBitHeight);
+      if (constrainingBitWidth <= SHRT_MAX)
```
remove space between `if` and `(`

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -93,25 +133,104 @@ ULONG TextFitNew(struct RastPort *rp, const char *string, ULONG strLen, struct T
+      if (constrainingBitWidth <= SHRT_MAX)
+      {
+        ULONG l_chars;
+        if(strLen0 != 0)
```
add empty line between variable definition and if()

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -93,25 +133,104 @@ ULONG TextFitNew(struct RastPort *rp, const char *string, ULONG strLen, struct T
+          tptr0   += l_chars;
+          result  += l_chars;
+        }
+        if (strLen0 > 0)
```
add empty line between } and if()

jens-maus commented on mcc/NewGfx.c:
```diff
@@ -93,25 +133,104 @@ ULONG TextFitNew(struct RastPort *rp, const char *string, ULONG strLen, struct T
         break;
 
+      constrainingBitWidth -= tabSizePixels;
+      te_width             += tabSizePixels;
```
remove spaces between `te_width` and `+= tabSizePixels;` no need to ident here.

alpyre replied:
I find them easier to read that way... but whatever... I can live without it. :)

jens-maus commented on mcc/PrintLineWithStyles.c:
```diff
@@ -254,12 +261,18 @@ LONG PrintLine(struct InstData *data, LONG x, struct line_node *line, LONG line_
+      clr_left  = dleft;
+      clr_width = MIN(flow+blockstart - data->xpos, (ULONG)_mwidth(data->object));
+                  
+      if (clr_width > 0)
```
remove space between `if` and `(`

jens-maus commented on mcc/PrintLineWithStyles.c:
```diff
@@ -286,55 +299,55 @@ LONG PrintLine(struct InstData *data, LONG x, struct line_node *line, LONG line_
-          if(cursor == TRUE &&
-             isFlagClear(data->flags, FLG_Active) &&
-             isFlagClear(data->flags, FLG_Activated))
+          if (left < dright || right > dleft)
```
remove space between `if` and `(`

jens-maus commented on mcc/PrintLineWithStyles.c:
```diff
@@ -286,55 +299,55 @@ LONG PrintLine(struct InstData *data, LONG x, struct line_node *line, LONG line_
 
+                       
+      {                  
+        LONG  x_start = MAX(dleft, (blockwidth ? xoffset+blockstart+blockwidth+flow : xoffset+blockstart+blockwidth)),
```
now that you touch these lines please add separate variable definition on each line and don't concatenate them via ','.

jens-maus commented on mcc/PrintLineWithStyles.c:
```diff
@@ -286,55 +299,55 @@ LONG PrintLine(struct InstData *data, LONG x, struct line_node *line, LONG line_
         }
-
-        DoMethod(data->object, MUIM_DrawBackground, x_start, y_start, x_width, y_width, x_ptrn, y_ptrn, 0);
+        if (x_start < dright)
```
remove space between `if` and `(`

jens-maus commented on mcc/PrintLineWithStyles.c:
```diff
@@ -391,28 +408,47 @@ LONG PrintLine(struct InstData *data, LONG x, struct line_node *line, LONG line_
 
-        if(fitting > 0)
+        // calculate how many chars will be ommited from start because of xpos
+        if (o_width && pen_pos < dleft)
```
remove space between `if` and `(`

jens-maus commented on mcc/PrintLineWithStyles.c:
```diff
@@ -491,26 +529,26 @@ LONG PrintLine(struct InstData *data, LONG x, struct line_node *line, LONG line_
       if(line_nr == 1)
       {
-        BltBitMapRastPort(rp->BitMap, xoffset, _mtop(data->object)-data->ypos, data->rport, _mleft(data->object), _mtop(data->object)+(data->fontheight * (line_nr-1)), _mwidth(data->object), data->fontheight-(_mtop(data->object)-data->ypos), (ABC|ABNC));
+        BltBitMapRastPort(rp->BitMap, 0, _mtop(data->object)-data->ypos, data->rport, _mleft(data->object), _mtop(data->object)+(data->fontheight * (line_nr-1)), _mwidth(data->object), data->fontheight-(_mtop(data->object)-data->ypos), (ABC|ABNC));
```
Please explain why in all of the following BltBitMapRaspPort() calls no `xoffset` is required anymore? Why was it necessary to change this logic?

alpyre replied:
Actually using xoffset value there was a **very very wrong** coding to begin with.

The blit copy from BitMap is **ONLY** done when in **doublebuffer** mode.
And when in doublebuffer mode, the **xoffset** value is always **0** (please see the lines **140** and **168** of **mcc/PrintLineWithStyles.c** in your repository) which makes sense because we should always be copying from the very start of the buffering bitmap. 

It should have been 0 to begin with.

jens-maus commented on mcc/PrintLineWithStyles.c:
```diff
@@ -286,55 +299,55 @@ LONG PrintLine(struct InstData *data, LONG x, struct line_node *line, LONG line_
       }
 
-
+                       
```
Remove empty line

jens-maus commented on mcc/Dispatcher.c:
```diff
@@ -956,6 +1008,23 @@ DISPATCHER(_Dispatcher)
                 * data->fontheight,
               MUIA_TextEditor_Prop_First, (data->visual_y-1)*data->fontheight,
               TAG_DONE);
+    // visual_y is changed it is best to recalculate longest visible line here
```
add empty line between `TAG_DONE);` and the comment on the next line.

jens-maus commented on mcc/PrintLineWithStyles.c:
```diff
@@ -344,11 +357,15 @@ LONG PrintLine(struct InstData *data, LONG x, struct line_node *line, LONG line_
+    o_width  = ((ULONG)flow > data->xpos ? 0 : data->xpos - flow);
+    maxwidth = data->WrapMode == MUIV_TextEditor_WrapMode_NoWrap ?
+              (MIN((_mwidth(data->object) + data->xpos - flow + rp->TxWidth), (ULONG)TextLengthNew(rp, text+x, c_length, data->TabSizePixels))) :
+                    _mwidth(data->object) - flow; 
```
add empty line after `flow;` and the next `while()`

jens-maus commented:
Thanks for all your patience. I have now successfully merged your changes (after having squashed them into a single commit) to the master branch of TextEditor.mcc. Next would be to actually test it throughoutly and then prepare a new release soon.

alpyre commented:
You are all wellcome. 
Thanks for the merge. It was tiresome but fun and very educative for me at the same time.
I'll pay more attention to code styles in my future commits.
...and before I forget...
...so many thanks for all that you've done for the Amiga platform through all those years.
